### PR TITLE
feat: Ar 2615 add flag in clients table to marks zombie clients and ignore while sending a message

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

when creating a session for a new client, the BE may response with null for the prekey, those clients are zombie clients that have to be excluded when sending messages.

### Solutions

1. add a flag `is_valid` to the clients table and mark null prekey client as invalid.
2. ignore invalid clients when encrypting messages to send.
3. clean up the code a bit.

### Testing

unit tests are added

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/CryptoSessionMapperTest.kt [hunk 1]
```diff
@@ -63,8 +47,14 @@ class CryptoSessionMapperTest {
             )
         )
 
-        cryptoSessionMapper.getMapOfSessionIdsToPreKeysAndIgnoreNull(qualifiedUserPreKeyInfo).also { actual ->
-            assertEquals(expected, actual)
+        val expectedInvalid: List<Pair<QualifiedIDEntity, List<String>>> = listOf(
+            UserIDEntity("user1", "domain1") to listOf("client1_null"),
+            UserIDEntity("user2", "domain1") to listOf("client1_null", "client2_null"),
+
+            )
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/CryptoSessionMapperTest.kt:54]:
```suggestion
        val expectedInvalid: List<Pair<QualifiedIDEntity, List<String>>> = listOf(
            UserIDEntity("user1", "domain1") to listOf("client1_null"),
            UserIDEntity("user2", "domain1") to listOf("client1_null", "client2_null"),
        )
```

Nitpick

**Review by vitorhugods — APPROVED** [review]
Sweet, sweet.

Just a couple of tiny suggestions. Feel free to ignore/leave for later.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt [hunk 2]
```diff
@@ -457,7 +457,7 @@ internal class ConversationDataSource internal constructor(
     override suspend fun getConversationRecipients(conversationId: ConversationId): Either<CoreFailure, List<Recipient>> =
         wrapStorageRequest {
             memberMapper.fromMapOfClientsEntityToRecipients(
-                clientDAO.getClientsOfConversation(idMapper.toDaoModel(conversationId))
+                clientDAO.conversationRepents(idMapper.toDaoModel(conversationId))
```
Comment (MohammedMokresh) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:460]: typo

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/prekey/PreKeyRepository.kt [hunk 3]
```diff
@@ -10,17 +10,17 @@ import com.wire.kalium.logic.data.user.UserId
...
 import com.wire.kalium.logic.wrapStorageRequest
 import com.wire.kalium.network.api.base.authenticated.prekey.PreKeyApi
+import com.wire.kalium.network.api.base.authenticated.prekey.PreKeyDTO
 import com.wire.kalium.persistence.dao.PrekeyDAO
 
 interface PreKeyRepository {
     suspend fun preKeysOfClientsByQualifiedUsers(
         qualifiedIdsMap: Map<UserId, List<ClientId>>
-    ): Either<NetworkFailure, List<QualifiedUserPreKeyInfo>>
+    ): Either<NetworkFailure, Map<String, Map<String, Map<String, PreKeyDTO?>>>>
```
Comment (MohammedMokresh) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/prekey/PreKeyRepository.kt:23]: maybe we can use type alias here to make it more readable

**Review by MohammedMokresh — APPROVED** [review]
just added few comments for naming and typos, but all looks good 🚀

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt [hunk 4]
```diff
@@ -98,25 +108,45 @@ internal class SessionEstablisherImpl internal constructor(
...
+                        if (prekey == null) {
+                            invalidList.add(QualifiedIDEntity(userId, domain) to clientId)
+                            false
+                        } else {
+                            true
+                        }
+                    }.mapValues {
+                        preKeyMapper.fromPreKeyDTO(it.value!!)
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (MohamadJaara) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt]: partition is for lists only, the preKeyInfo here is a Map i guess the naming need to be updated 😆

**Pull request merged**